### PR TITLE
Created for discussion of errors and getting advises

### PR DESCRIPTION
**Pull request by dmytrokohut**

### CODE REVIEW

File: src/main/java/com/example/springboot/Base/Book.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+		this.title = title;
+		this.author = author;
+		this.readerId = readerId;
+	}
+	
+	/**
+	 * This setter sets a the value of 'id' attribute
+	 * @param id
+	 */
+	public void setId(int id) {
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Base/Book.java]: No need to write docs for set and get

File: src/main/java/com/example/springboot/Base/Book.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	 */
+	public int getReaderId() {
+		return readerId;
+	}
+	
+	/**
+	 * The method which represent a object in string view
+	 * @return
+	 */
+	@Override
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Base/Book.java]: Why do you need this method?

File: src/main/java/com/example/springboot/Base/Reader.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
+package com.example.springboot.Base;
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Base/Reader.java]: The package name must be a lowercase

File: src/main/java/com/example/springboot/Base/Reader.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+		this.id = id;
+		this.name = name;
+		this.email = email;
+	}
+	
+	/**
+	 * This setter-method sets the value of 'id' attribute
+	 * @param id
+	 */
+	public void setId(int id) {
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Base/Reader.java]: No need to write docs for set and get

File: src/main/java/com/example/springboot/Base/Reader.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	public String getEmail() {
+		return email;
+	}
+	
+	/**
+	 * This method return a string representation of object
+	 * @return
+	 */
+	@Override
+	public String toString() {
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Base/Reader.java]: Why do you need this method?

File: src/main/java/com/example/springboot/Controller/BookController.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
+package com.example.springboot.Controller;
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Controller/BookController.java]: The package name must be a lowercase

File: src/main/java/com/example/springboot/Controller/BookController.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+import java.util.Collection;
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Controller/BookController.java]: Do you use this import?

File: src/main/java/com/example/springboot/Controller/BookController.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		return bookDAO.findAllBooks();
+	}
+	
+	/**
+	 * Looking for a book with given id
+	 * @param id
+	 * @return 
+	 */
+	@RequestMapping(value="/select/{id}", method=RequestMethod.GET)
+	public Book selectById(@PathVariable("id") int id) {
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Controller/BookController.java]: Better to use Integer

File: src/main/java/com/example/springboot/Controller/BookController.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+		return bookDAO.selectById(id);
+	}
+	
+	/**
+	 * Create a new book in database
+	 * @param book
+	 * @return
+	 */
+	@RequestMapping(value="/create", method=RequestMethod.POST)
+	public String createBook(@RequestBody Book book) {
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Controller/BookController.java]: bad idea to use String in response

File: src/main/java/com/example/springboot/Controller/BookController.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+		return "New book was created !";
+	}
+	
+	/**
+	 * Update information about book
+	 * @param book
+	 * @return
+	 */
+	@RequestMapping(value="/update", method=RequestMethod.PUT)
+	public String updateBook(@RequestBody Book book) {
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Controller/BookController.java]: ad idea to use String in response

Comment (ZVIproject) [reply src/main/java/com/example/springboot/Controller/BookController.java]: try to use JSON

File: src/main/java/com/example/springboot/Controller/BookController.java (hunk 9 above)
Comment (ZVIproject) [reply src/main/java/com/example/springboot/Controller/BookController.java]: try to use JSON

Comment (ZVIproject) [reply src/main/java/com/example/springboot/Controller/BookController.java]: and method name should be create, why createBook, the name of the class is a BookController

File: src/main/java/com/example/springboot/Controller/ReaderController.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+ * This is class that control work of application, process requests where URL start with '/reader'
+ * @author Dmytro Kohut
+ * @version 1.0
+ */
+@RestController
+@RequestMapping("/reader")
+public class ReaderController {
+	
+	@Autowired
+	private ReaderDAO readerDAO;
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Controller/ReaderController.java]: Interface

File: src/main/java/com/example/springboot/Controller/ReaderController.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
+package com.example.springboot.Controller;
+
+import java.util.Collection;
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Controller/ReaderController.java]: Do you use it?

File: src/main/java/com/example/springboot/Controller/ReaderController.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
+package com.example.springboot.Controller;
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Controller/ReaderController.java]: package name

File: src/main/java/com/example/springboot/Controller/ReaderController.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+		return readerDAO.selectById(id);
+	}
+	
+	/**
+	 * Create a new reader in database
+	 * @param book
+	 * @return
+	 */
+	@RequestMapping(value="/create", method=RequestMethod.POST)
+	public String createReader(@RequestBody Reader reader){
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Controller/ReaderController.java]: class name Reader

File: src/main/java/com/example/springboot/Controller/ReaderController.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+		return "Reader - was created";
+	}
+	
+	/**
+	 * Update information about reader
+	 * @param book
+	 * @return
+	 */
+	@RequestMapping(value="/update", method=RequestMethod.PUT)
+	public String updateReader(@RequestBody Reader reader) {
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Controller/ReaderController.java]: class name Reader

File: src/main/java/com/example/springboot/Controller/ReaderController.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+		return "Reader with id=" + reader.getId() + " was updated";
+	}
+	
+	/**
+	 * Delete reader from database
+	 * @param id
+	 * @return
+	 */
+	@RequestMapping(value="/delete/{id}", method=RequestMethod.DELETE)
+	public String deleteReader(@PathVariable("id") int id) {
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/Controller/ReaderController.java]: class name Reader

File: src/main/java/com/example/springboot/DAO/BookDAO.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
+package com.example.springboot.DAO;
+
+import java.util.Collection;
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/DAO/BookDAO.java]: Why are you always using this import?)

File: src/main/java/com/example/springboot/DAO/BookDAO.java [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.example.springboot.Base.Book;
+
+/**
+ * This interface defines a methods for work with database table - "Book"
+ * @author Dmytro Kohut
+ * @version 1.0
+ */
+public interface BookDAO {
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/DAO/BookDAO.java]: You can create one interface for book and Reader for this operations, but use a templates

File: src/main/java/com/example/springboot/DAO/BookDAOImpl.java [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+	
+	/**
+	 * @see com.example.springboot.DAO.BookDAO#findAllBooks()
+	 */
+	@Override
+	public Collection<Book> findAllBooks() {
+		Connection connection = null;
+		List<Book> resultList = new ArrayList<>();
+		
+		try {
```
Comment (ZVIproject) [inline src/main/java/com/example/springboot/DAO/BookDAOImpl.java]: Why you catch it?

File: src/main/java/com/example/springboot/base/Book.java [hunk 20]
```diff
@@ -0,0 +1,68 @@
+package com.example.springboot.base;
+
+/**
+ * This is a basic class for book object.
+ * @author Dmytro Kohut
+ * @version 1.1
+ */
+public class Book {
+	public Integer id;
```
Comment (koropatva) [inline src/main/java/com/example/springboot/base/Book.java:9]: `public` really?

File: src/main/java/com/example/springboot/dao/ReaderDAOImpl.java [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+	 */
+	private Connection getConnection() throws SQLException {
+		return DriverManager.getConnection("jdbc:mysql://localhost:3306/libriary", "root", "root");
+	}
+	
+	/**
+	 * This method close a given connection
+	 * @param connection
+	 */
+	private void closeConnection(Connection connection) {
```
Comment (koropatva) [inline src/main/java/com/example/springboot/dao/ReaderDAOImpl.java:49]: you are using java 7 or java 6? on java 7+ connections can be closed automatic

File: src/main/java/com/example/springboot/dao/ReaderDAOImpl.java [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+		}
+	}
+	
+	/**
+	 * This method create and return a connection for database access
+	 * @return
+	 * @throws SQLException 
+	 */
+	private Connection getConnection() throws SQLException {
+		return DriverManager.getConnection("jdbc:mysql://localhost:3306/libriary", "root", "root");
```
Comment (koropatva) [inline src/main/java/com/example/springboot/dao/ReaderDAOImpl.java:42]: WTF? why do you need spring properties if you are hardcoded?

File: src/main/java/com/example/springboot/dao/DAOInterface.java [hunk 23]
```diff
@@ -0,0 +1,42 @@
+package com.example.springboot.dao;
+
+import java.util.Collection;
+
+/**
+ * This interface defines a methods for work with database
+ * @author Dmytro Kohut
+ * @version 1.1
+ */
+public interface DAOInterface<T> {
```
Comment (koropatva) [inline src/main/java/com/example/springboot/dao/DAOInterface.java:10]: `DAOInterface` really? WTF? You have to create new interface for each DAO, not one abstract for all of them

File: src/main/java/com/example/springboot/dao/BookDAOImpl.java [hunk 24]
```diff
@@ -0,0 +1,181 @@
...
+import com.example.springboot.base.Book;
+import com.mysql.jdbc.Statement;
+
+/**
+ * This class implement methods from BookDAO interface
+ * @author Dmytro Kohut
+ * @version 1.1
+ */
+@Repository("bookDAO")
+public class BookDAOImpl implements DAOInterface<Book> {
```
Comment (koropatva) [inline src/main/java/com/example/springboot/dao/BookDAOImpl.java:24]: I don't like when services calls `Impl`. Rewrite it BookDAOImpl have to be `BookDAOService` and `DAOInterface` -> `IBookDAOService`

File: src/main/java/com/example/springboot/controller/ReaderController.java [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+	
+	static final private ObjectMapper objectMapper = new ObjectMapper();
+	
+	/**
+	 * Get information about all readers
+	 * @return
+	 */
+	@RequestMapping(value="/all", method=RequestMethod.GET)
+	public Collection<Reader> findAllReaders(){
+		return readerDAO.findAll();
```
Comment (koropatva) [inline src/main/java/com/example/springboot/controller/ReaderController.java:39]: WTF? You have started using Spring REST, so please use Spring response classes for it

File: src/main/java/com/example/springboot/controller/ReaderController.java [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+	private DAOInterface<Reader> readerDAO;
+	
+	static final private ObjectMapper objectMapper = new ObjectMapper();
+	
+	/**
+	 * Get information about all readers
+	 * @return
+	 */
+	@RequestMapping(value="/all", method=RequestMethod.GET)
+	public Collection<Reader> findAllReaders(){
```
Comment (koropatva) [inline src/main/java/com/example/springboot/controller/ReaderController.java:38]: if it is controller for `Reader` it means method `findAll` will return `Readers` why do you need remember it on method name?

File: src/main/java/com/example/springboot/controller/BookController.java [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+	
+	/**
+	 * Delete book in database by given id
+	 * @param id
+	 * @return
+	 * @throws JsonProcessingException 
+	 * @throws IOException 
+	 */
+	@RequestMapping(value="/delete/{id}", method=RequestMethod.DELETE)
+	public JsonNode delete(@PathVariable("id") Integer id) throws JsonProcessingException, IOException {
```
Comment (koropatva) [inline src/main/java/com/example/springboot/controller/BookController.java:87]: Return `JsonNode`? Are you kidding me

File: src/main/java/com/example/springboot/dao/BookDAOImpl.java [hunk 28]
```diff
@@ -0,0 +1,181 @@
...
+
+import com.example.springboot.base.Book;
+import com.mysql.jdbc.Statement;
+
+/**
+ * This class implement methods from BookDAO interface
+ * @author Dmytro Kohut
+ * @version 1.1
+ */
+@Repository("bookDAO")
```
Comment (koropatva) [inline src/main/java/com/example/springboot/dao/BookDAOImpl.java:23]: WTF? DON'T REWRITE IMPLEMENTATION NAMES!!! NEVER EVER DO IT!!!!

File: src/main/java/com/example/springboot/dao/BookDAOImpl.java [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+public class BookDAOImpl implements DAOInterface<Book> {
+	
+	final static Logger LOGGER = Logger.getLogger(BookDAOImpl.class);
+	
+	static {
```
Comment (koropatva) [inline src/main/java/com/example/springboot/dao/BookDAOImpl.java:28]: What are hell?

File: src/main/java/com/example/springboot/dao/BookDAOImpl.java [hunk 30]
```diff
@@ -0,0 +1,181 @@
...
+	 * @see com.example.springboot.dao.DAOInterface#findAllBooks()
+	 */
+	@Override
+	public Collection<Book> findAll() {
+		Connection connection = null;
+		List<Book> resultList = new ArrayList<>();
+		
+		try {
+			connection = getConnection();
+			PreparedStatement statement = connection.prepareStatement("SELECT * FROM `books`;");
```
Comment (koropatva) [inline src/main/java/com/example/springboot/dao/BookDAOImpl.java:67]: Move query into static variable

Comment (koropatva) [reply src/main/java/com/example/springboot/dao/BookDAOImpl.java:67]: On the query you HAVE TO use list of fields, that you are trying to fetch not `*`

File: src/main/java/com/example/springboot/dao/BookDAOImpl.java [hunk 31]
```diff
@@ -0,0 +1,181 @@
...
+		}
+	}
+	
+	/**
+	 * @see com.example.springboot.dao.DAOInterface#selectById(int id)
+	 */
+	@Override
+	public Book selectById(Integer id) {
+		Connection connection = null;
+		Book book = new Book();
```
Comment (koropatva) [inline src/main/java/com/example/springboot/dao/BookDAOImpl.java:117]: WTF?

File: src/main/resources/application.properties [hunk 32]
```diff
@@ -0,0 +1,4 @@
+spring.datasource.url=jdbc:mysql://localhost:3306/libriary
```
Comment (koropatva) [inline src/main/resources/application.properties:1]: why do you need properties, that you never used????

File: src/main/java/com/example/springboot/DAO/BookDAO.java (hunk 18 above)
Comment (koropatva) [reply src/main/java/com/example/springboot/DAO/BookDAO.java]: or my God... One interface for all services. WTF??? Are you kidding me?

**Pull request closed without merging**